File: examples3d/primitives3.rs
use na::Point3;
use rapier3d::dynamics::{JointSet, RigidBodyBuilder, RigidBodySet};
use rapier3d::geometry::{ColliderBuilder, ColliderSet};
use rapier_testbed3d::Testbed;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let joints = JointSet::new();

    /*
     * Create the cubes
     */
    let num = 8;
    let rad = 1.0;

    let shiftx = rad * 2.0 + rad;
    let shifty = rad * 2.0 + rad;
    let shiftz = rad * 2.0 + rad;
    let centerx = shiftx * (num / 2) as f32;
    let centery = shifty / 2.0;
    let centerz = shiftz * (num / 2) as f32;

    let mut offset = -(num as f32) * (rad * 2.0 + rad) * 0.5;

    for j in 0usize..1 {
        for i in 0..1 {
            for k in 0usize..1 {
                let x = i as f32 * shiftx - centerx + offset;
                let y = j as f32 * shifty + centery - rad;
                let z = k as f32 * shiftz - centerz + offset;

                // Build the rigid body.
                let rigid_body = RigidBodyBuilder::new_dynamic().translation(x, y, z).build();
                let handle = bodies.insert(rigid_body);

                let collider = match j % 5 {
                    0 => ColliderBuilder::cuboid(rad, rad, rad).build(),
                    1 => ColliderBuilder::ball(rad).build(),
                    // Rounded cylinders are much more efficient that cylinder, even if the
                    // rounding margin is small.
                    2 => ColliderBuilder::round_cylinder(rad, rad, rad / 10.0).build(),
                    3 => ColliderBuilder::cone(rad, rad).build(),
                    _ => ColliderBuilder::capsule_y(rad, rad).build(),
                };

                colliders.insert(collider, handle, &mut bodies);
            }
        }

        offset -= 0.05 * rad * (num as f32 - 1.0);
    }

    /*
     * Ground
     */
    testbed.add_callback(
        move |mut window, mut graphics, physics, events, run_state| {
            if run_state.timestep_id == 10 {
                let ground_size = 100.1;
                let ground_height = 2.1;

                let rigid_body = RigidBodyBuilder::new_static()
                    .translation(0.0, -ground_height, 0.0)
                    .build();
                let handle = physics.bodies.insert(rigid_body);
                let collider =
                    ColliderBuilder::cuboid(ground_size, ground_height, ground_size).build();
                physics
                    .colliders
                    .insert(collider, handle, &mut physics.bodies);

                if let (Some(graphics), Some(window)) = (&mut graphics, &mut window) {
                    graphics.add(window, handle, &physics.bodies, &physics.colliders);
                }
            }
        },
    );

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, joints);
    testbed.physics_state_mut().gravity.fill(0.0);
    testbed.look_at(Point3::new(100.0, 100.0, 100.0), Point3::origin());
}

fn main() {
    let testbed = Testbed::from_builders(0, vec![("Boxes", init_world)]);
    testbed.run()
}

File: python/rapier-testbed/rapier_testbed/examples3/primitives3.py
"""Port of ``examples3d/primitives3.rs``.

A grid of 8 x 8 x 20 mixed primitives (ball, round_cylinder, cone,
capsule_y) falling onto a wide ground cuboid. Same shape selection rule as
the Rust example: ``j % 5`` cycles primitive kind every vertical layer.
"""

from __future__ import annotations

import rapier3d as rp

from .._registry import register


CATEGORY = "Collisions"
NAME = "Primitives"


def init_world(testbed) -> None:  # type: ignore[no-untyped-def]
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    # Ground.
    ground_size = 100.1
    ground_height = 2.1

    ground_body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_handle = bodies.insert(ground_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size),
        ground_handle,
        bodies,
    )

    # Falling primitive grid.
    num = 8
    rad = 1.0
    shiftx = rad * 2.0 + rad
    shifty = rad * 2.0 + rad
    shiftz = rad * 2.0 + rad
    centerx = shiftx * (num // 2)
    centery = shifty / 2.0
    centerz = shiftz * (num // 2)

    offset = -float(num) * (rad * 2.0 + rad) * 0.5

    for j in range(20):
        for i in range(num):
            for k in range(num):
                x = i * shiftx - centerx + offset
                y = j * shifty + centery + 3.0
                z = k * shiftz - centerz + offset

                body = rp.RigidBody.dynamic().translation((x, y, z))
                handle = bodies.insert(body)

                kind = j % 5
                if kind == 1:
                    collider = rp.Collider.ball(rad)
                elif kind == 2:
                    # Rounded cylinder is much faster than cylinder even
                    # for a small rounding margin (matches the Rust note).
                    collider = rp.Collider.round_cylinder(rad, rad, rad / 10.0)
                elif kind == 3:
                    collider = rp.Collider.cone(rad, rad)
                else:
                    collider = rp.Collider.capsule_y(rad, rad)

                colliders.insert_with_parent(collider, handle, bodies)

        offset -= 0.05 * rad * (num - 1)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


# Register at import time so ``rapier_testbed.EXAMPLES`` sees us.
register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    # ``python -m rapier_testbed.examples3.primitives3``
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
